fdbserver/DataDistributionQueue.actor.cpp

```diff
@@ -1807,20 +1958,171 @@ ACTOR Future<bool> rebalanceReadLoad(DDQueueData* self,
+		    // shards.size());
+		    shards,
+		    K,
+		    srcLoad);
+		req.comparator = [](const StorageMetrics& a, const StorageMetrics& b) {
+			return a.bytesReadPerKSecond / std::max(a.bytes * 1.0, 1.0) >
+			       b.bytesReadPerKSecond / std::max(b.bytes * 1.0, 1.0);
+		};
+		GetTopKMetricsReply reply = wait(brokenPromiseToNever(self->getTopKMetrics.getReply(req)));
+		return reply;
+	} else { // get BottomK shards
+		GetTopKMetricsRequest req(shards, K, srcLoad / shards.size());
+		req.comparator = [](const StorageMetrics& a, const StorageMetrics& b) {
+			return a.bytesReadPerKSecond / std::max(a.bytes * 1.0, 1.0) <
+			       b.bytesReadPerKSecond / std::max(b.bytes * 1.0, 1.0);
+		};
+
+		GetTopKMetricsReply reply = wait(brokenPromiseToNever(self->getTopKMetrics.getReply(req)));
+		return reply;
+	}
+}
+
+// TODO@ZZX: reduce replicas case
+RelocateOperation selectRelocateOperation(DDQueueData* self,
+                                          TraceEvent* traceEvent,
+                                          DataMovementReason reason,
+                                          double minReadLoad,
+                                          double maxReadLoad,
+                                          double srcLoad,
+                                          StorageMetrics& metrics,
+                                          GetTopKMetricsReply& bottomKReply) {
+	RelocateOperation operation = RelocateOperation::INVALID;
+	double epsilon = 0.0001;
+	int64_t selectShardReadLoad = metrics.bytesReadPerKSecond;
+
+	std::pair<std::vector<ShardsAffectedByTeamFailure::Team>, std::vector<ShardsAffectedByTeamFailure::Team>> teams =
+	    self->shardsAffectedByTeamFailure->getTeamsFor(metrics.keys.get());
+	int srcTeamSize = teams.first.size();
+	if (!(SERVER_KNOBS->DYNAMIC_REPLICATION_ENABLED)) {
+		// If the button was turned off, we directly use old ways, i.e, move shards to dest team
+		operation = RelocateOperation::MOVE;
+	} else if (isDataMovementForMountainChopper(reason)) {
+		if (srcTeamSize > self->teamSize) {
+			// Already have an auxiliary team
+			operation = RelocateOperation::INVALID;
+		} else if (fabs(minReadLoad - maxReadLoad) < epsilon && fabs(minReadLoad - selectShardReadLoad) < epsilon) {
+			// If three of them are equal, it means we only have one shard whose readLoad > 0 in sourceTeam. In this
+			// case, moving shard to another team won't solve the problem, so we would use MULTIPLY_REPLICAS here.
+			operation = RelocateOperation::MULTIPLY_REPLICAS;
+		} else if (selectShardReadLoad >= srcLoad * SERVER_KNOBS->DYNAMIC_REPLICATION_SHARD_BANDWIDTH_FRAC) {
+			// If one team has a very hot shard, we will try to multiply replicas.
+			operation = RelocateOperation::MULTIPLY_REPLICAS;
+		} else {
+			operation = RelocateOperation::MOVE;
+		}
+	} else {
+		// ValleyFiller will check bottomK(REDUCE_REPLICAS) and TopK(MOVE) metrics together.
```

Comment:
Is this coming from ValleyFiller? We were planning a separate actor that would randomly select teams, correct?

Reply:
Now, both MountainChopper & ValleyFiller would check reduceShard case. We may want to create a new actor in parallel to those two in the future.